sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs

```diff
@@ -1949,6 +1949,147 @@ private void SetNameFieldsIfNull()
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response"/>.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs
+        /// </remarks>
+        public virtual async Task<Response<ShareInfo>> RestoreAsync(
+            string deletedShareName,
+            string deletedShareVersion,
+            CancellationToken cancellationToken = default)
+            => await RestoreInternal(
+                deletedShareName,
+                deletedShareVersion,
+                async: true,
+                cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Restores a previously created share.  The restored share
+        /// will be renamed to the name of this <see cref="ShareServiceClient"/>.
+        /// If the container associated with this <see cref="ShareServiceClient"/>
+        /// already exists, this call will result in a 409 (conflict).
+        /// This API is only functional is Share Soft Delete is enabled
+        /// for the storage account associated with the share.
+        /// </summary>
+        /// <param name="deletedShareName">
+        /// The name of the share to restore.
+        /// </param>
+        /// <param name="deletedShareVersion">
+        /// The version of the share to restore.
+        /// </param>
+        /// <param name="async">
+        /// Whether to invoke the operation asynchronously.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response"/>.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        private async Task<Response<ShareInfo>> RestoreInternal(
+            string deletedShareName,
+            string deletedShareVersion,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            using (Pipeline.BeginLoggingScope(nameof(ShareServiceClient)))
+            {
+                Pipeline.LogMethodEnter(
+                    nameof(ShareServiceClient),
+                    message:
+                    $"{nameof(Uri)}: {Uri}\n" +
+                    $"{nameof(deletedShareName)}: {deletedShareName}\n" +
+                    $"{nameof(deletedShareVersion)}: {deletedShareVersion}");
+
+                try
+                {
+                    return await FileRestClient.Share.RestoreAsync(
+                        ClientDiagnostics,
+                        Pipeline,
+                        Uri,
+                        Version.ToVersionString(),
```

Comment:
Does the operation just fail if you call this with an older service version?  I can't imagine this causing any problems, but want to double check we don't need a version guard here.

Reply:
Yes it will fail, with a message from the service saying the version is incorrect.  Peter decided we weren't doing version guarding in the SDK.

Reply:
I thought it was specifically no guards unless it could lead to data corruption or loss - which is why I like double checking for some of these new operations.

Reply:
Nothing will happen to your share if you call Restore with a service version that doesn't support it - you will just get a 400 class error.